src/coord/src/coord.rs

```diff
@@ -588,23 +588,22 @@ impl Coordinator {
         }
     }
 
-    fn message_worker(
-        &mut self,
-        dataflow::Response {
-            worker_id: _,
-            message,
-        }: dataflow::Response,
-    ) {
+    fn message_worker(&mut self, dataflow::Response { worker_id, message }: dataflow::Response) {
         match message {
             WorkerFeedback::PeekResponse(conn_id, response) => {
                 // We use an `if let` here because the peek could have been cancelled already.
-                if let Some((channel, countdown)) = self.pending_peeks.get_mut(&conn_id) {
-                    channel
-                        .send(response)
-                        .expect("Peek endpoint terminated prematurely");
-                    *countdown -= 1;
-                    if *countdown == 0 {
-                        self.pending_peeks.remove(&conn_id);
+                if let Some((channel, workers_responded)) = self.pending_peeks.get_mut(&conn_id) {
+                    // Only accept responses for workers we haven't heard from yet.
+                    if workers_responded.insert(worker_id) {
```

Comment:
Should we panic or log in the false case? If that should never happen, it should maybe be enforced? I'm really not sure.

Reply:
If it ever does trigger, that perhaps should mean we don't know what to do because we can no longer trust dataflow. Our current assumption is that in order to provide correct answers we get exactly one response per worker. If we ever don't, then we can no longer be assured that the answers we are providing to the user are correct.

Reply:
I think it might come down to a different change I saw, where peeks might have a nonce with them so that it is unambiguous what a peek response is responding to. If that is in place, it seems like one could more easily log the error and reason that future peeks with new nonces won't be negatively affected. If that change is just about to come down the pipe, maybe no worries sorting out the nuance of this case in the meantime?

Reply:
I'm also ok with, say, an assert that panics in dev but logs in prod? Is that a thing or a good idea? I'm fine moving forward with just a log too if that allows things to proceed.

Reply:
I support that (we .. have some common code somewhere via @benesch I think for asserting in CI, essentially). But I think that pattern is too weak for situations where we have really lost confidence in the integrity of the connection (e.g., if it would mean that `coord` and `dataflow` have de-sync'd for some reason). Great reason to toss a nonce in there, whoever had that idea! (not me).

Reply:
I've converted these potentially loggable scenarios in panics just to a) see what happens b) merge if it seems OK.

@frankmcsherry The nonce was a speculative design to figure out how easily we could provide the same invariant as sending a cancel response without waiting for the dataflow to respond. If we brought the nonce design to bear, would you suggest sending an error to the channel on encountering an out of sync nonce, i.e. surface the error to clients in hopes that it's temporary?

Reply:
I was mostly thinking if there were nonces, we wouldn't have to set MZ on fire, because we could reasonably believe we would "resync" with the next Peek, and the coordinator could just log "dataflow seems to have screwed up peek xyz" instead. I'm not sure if you should error at the consumer, tbh.